src/lib.rs

```diff
@@ -1,21 +1,392 @@
 //! The project that will henceforth be known as `starling`.
-
+//!
+// `error_chain!` can recurse deeply
+#![recursion_limit = "1024"]
 #![deny(missing_docs)]
 #![deny(missing_debug_implementations)]
+#![deny(unsafe_code)]
+// Annoying warning emitted from the `error_chain!` macro.
+#![allow(unused_doc_comment)]
 
+#[macro_use]
+extern crate derive_is_enum_variant;
+#[macro_use]
+extern crate error_chain;
+#[macro_use]
 extern crate futures;
+extern crate futures_cpupool;
+extern crate futures_mpsc;
+#[macro_use]
+extern crate js;
+extern crate num_cpus;
 extern crate tokio_core;
 
-use futures::future;
-use std::io;
-use tokio_core::reactor::Core;
+pub(crate) mod task;
+
+use futures::{future, Future, Sink, Stream};
+use futures_cpupool::CpuPool;
+use futures_mpsc::{Receiver, Sender};
+use std::cmp;
+use std::collections::HashMap;
+use std::fmt;
+use std::mem;
+use std::path;
+use std::sync::Arc;
+use std::thread;
+
+error_chain! {
+    foreign_links {
+        Io(::std::io::Error)
+        /// An IO error.
+            ;
+
+        SendError(futures_mpsc::SendError<()>)
+        /// Tried to send a value on a channel when the receiving half was
+        /// already dropped.
+            ;
+    }
+
+    errors {
+        /// Could not create a JavaScript runtime.
+        CouldNotCreateJavaScriptRuntime {
+            description("Could not create a JavaScript Runtime")
+            display("Could not create a JavaScript Runtime")
+        }
+
+        /// Could not read a value from a channel.
+        CouldNotReadValueFromChannel {
+            description("Could not read a value from a channel")
+            display("Could not read a value from a channel")
+        }
+
+        /// There was an exception in JavaScript code.
+        // TODO: stack, line, column, filename, etc
+        JavaScriptException {
+            description("JavaScript exception")
+            display("JavaScript exception")
+        }
+    }
+}
+
+impl Clone for Error {
+    fn clone(&self) -> Self {
+        self.to_string().into()
+    }
+}
 
-/// Run the main `starling` loop.
-pub fn run() -> Result<(), io::Error> {
-    let mut core = Core::new()?;
-    core.run(future::ok(()))
+/// Configuration options for building a `starling` event loop.
+///
+/// ```
+/// extern crate starling;
+///
+/// # fn foo() -> starling::Result<()> {
+/// // Construct a new `Options` builder, providing the file containing
+/// // the main JavaScript task.
+/// starling::Options::new("path/to/main.js")
+///     // Finish configuring the `Options` builder and run the event
+///     // loop!
+///     .run()?;
+/// # Ok(())
+/// # }
+/// ```
+#[derive(Clone, Debug)]
+pub struct Options {
+    main: path::PathBuf,
+    sync_io_pool_threads: usize,
+    cpu_pool_threads: usize,
+    channel_buffer_size: usize,
 }
 
-#[test]
-fn it_works() {
+const DEFAULT_SYNC_IO_POOL_THREADS: usize = 8;
```

Comment:
The number of threads in the sync IO pool is more about space/IO throughput trade off than how many CPUs are available, since most of these threads are going to be waiting on IO rather than using the CPU.

Reply:
True. Ideally this won't be used much anyway - we should be able to find async versions of most I/O stuff eventually, I'd hope.